packages/join-block/src/Services/JoinService.php

```diff
@@ -23,16 +23,69 @@ private static function formatDateForChargebee($day, $month, $year)
 
     public static function handleJoin($data)
     {
+        $lockFile = null;
         try {
+            $sessionToken = $data['sessionToken'] ?? null;
+            $lockFile = self::lockSession($sessionToken);
             $chargebeeCustomer = self::tryHandleJoin($data);
             do_action('ck_join_flow_success', $data, $chargebeeCustomer);
         } catch (\Exception $e) {
             do_action('ck_join_flow_error', $data, $e);
             throw $e;
+        } finally {
+            self::unlockSession($lockFile);
         }
         return $chargebeeCustomer;
     }
 
+    /**
+     * This is a BLOCKING lock, which means threads will sleep
+     * until they can get the lock. This forces sequential execution,
+     * which means no race conditions.
+     *
+     * We still need to handle duplicate join requests, by
+     * e.g. making sure the code doesn't create a subscription
+     * if one already exists.
+     */
+    public static function lockSession($sessionToken)
+    {
+        global $joinBlockLog;
+
+        if (!$sessionToken) {
+            throw new \Exception("Unable to lock session: no token provided");
+        }
+
+        $joinBlockLog->info("Locking session $sessionToken");
+
+        $lockFilepath = sys_get_temp_dir() . '/' . $sessionToken;
```

Comment:
Is it possible that this will work on the shared hosting LRU have?

Reply:
Instead put this somewhere writable within the WordPress hierarchy, because we can't guarantee we have this directory available.